app/waf/src/controllers/adc.controller.ts

```diff
@@ -193,4 +207,143 @@ export class AdcController {
+            );
+
+          // TODO: Create VM in async way.
+          return await this.createOn(adc).then(async () => {
+            return {id: adc.id};
+          });
+        } catch (error) {
+          throw new HttpErrors.BadRequest(error.message);
+        }
+
+      case 'delete':
+        break;
+
+      case 'setup':
+        break;
+
+      default:
+        throw new HttpErrors.BadRequest(
+          'Not supported: ' + Object.keys(actionBody)[0],
+        );
+    }
+  }
+
+  private async serialize(adc: Adc, data?: object) {
+    if (data) Object.assign(adc, data);
+    await this.adcRepository.update(adc);
+  }
+
+  private async createOn(adc: Adc): Promise<void> {
+    try {
+      await this.serialize(adc, {status: 'BUILDING'})
+        .then(async () => await this.cNet(adc))
+        .then(async () => await this.cSvr(adc));
+      await this.serialize(adc, {status: 'POWERON'});
+    } catch (error) {
+      await this.serialize(adc, {status: 'ERROR'});
+      throw error;
+    }
+  }
+
+  private async cSvr(adc: Adc): Promise<void> {
+    await Promise.all([
+      this.wafapp.get(WafBindingKeys.KeyComputeManager),
+      this.reqCxt.get(WafBindingKeys.Request.KeyUserToken),
+      this.reqCxt.get(WafBindingKeys.Request.KeyTenantId),
+    ]).then(async ([computeHelper, userToken, tenantId]) => {
+      let serverParams: ServersParams = {
+        userTenantId: tenantId,
+        vmName: adc.id,
+        imageRef: adc.compute.imageRef,
+        flavorRef: adc.compute.flavorRef,
+        securityGroupName: 'default', //TODO: remove the hardcode in the future.
+        regionName: 'RegionOne',
+        ports: (() => {
+          let ports = [];
+          for (let n of Object.keys(adc.networks)) {
+            ports.push(<string>adc.networks[n].portId);
+          }
+          return ports;
+        })(),
+      };
+
+      await computeHelper
+        .createServer(userToken, serverParams)
+        .then(response => {
+          adc.compute.vmId = response;
+          adc.management = {
+            tcpPort: 443, // TODO: remove the hard-code.
+            ipAddress: <string>(() => {
+              let ip: string | undefined;
+              Object.keys(adc.networks).forEach(v => {
+                if (adc.networks[v].type === 'mgmt')
+                  return adc.networks[v].fixedIp;
+              });
+              return ip;
+            })(),
+          };
+        });
+    });
+  }
+
+  private async cNet(adc: Adc): Promise<void> {
+    await Promise.all([
+      this.wafapp.get(WafBindingKeys.KeyNetworkDriver),
+      this.reqCxt.get(WafBindingKeys.Request.KeyUserToken),
+    ]).then(async ([networkHelper, userToken]) => {
+      for (let k of Object.keys(adc.networks)) {
+        let net = adc.networks[k];
+        if (net.portId && net.ready) continue;
+
+        net.ready = false;
+
+        let portParams: PortCreationParams = {
+          networkId: net.networkId,
+          regionName: 'RegionOne',
```

Comment:
is regionName a must-have?  seems to me it is not needed.

Reply:
In the PR: https://github.com/F5Networks/f5-adcaas-openstack/pull/106, `regionName` was hidden. That should answer your(and my) expectation. 

In actual usage, after auth* with keystone, the endpoints are returned within token's `catalog` always tied with a region tag, like: 

```
              {
                region_id: 'RegionOne',
                url:
                  'http://localhost:8774/v2/fde45211da0a44ecbf38cb0b644ab30d',
                region: 'RegionOne',
                interface: 'public',
                id: 'c776745dbad64475a020e219b636894d',
              },
```

So in the above-mentioned PR, I hide it into the auth token handling(with a default value 'RegionOne'). Users can still appoint it if it's not 'RegionOne'.

Reply:
why would user appoint region information during create-port ?

Reply:
The region info is not needed for creating port. but needed to get the network endpoint. 

```
  async createPort(
    userToken: string,
    portParams: PortCreationParams,
  ): Promise<PortResponse> {
    try {
      // TODO: exception check.
      return await this.networkEndpoint(portParams.regionName)
        .then(async networkUrl => {
          let url = networkUrl + '/v2.0/ports';
          let body: PortsRequest = {
            port: {
              network_id: portParams.networkId,
            },
          };
```

It will be hidden soon. What's more, it's not appointed by user. 

See here: 
```
  private async cNet(adc: Adc): Promise<void> {
    await Promise.all([
      this.wafapp.get(WafBindingKeys.KeyNetworkDriver),
      this.reqCxt.get(WafBindingKeys.Request.KeyUserToken),
    ]).then(async ([networkHelper, userToken]) => {
      for (let k of Object.keys(adc.networks)) {
        let net = adc.networks[k];
        if (net.portId && net.ready) continue;

        net.ready = false;

        let portParams: PortCreationParams = {
          networkId: net.networkId,
          regionName: 'RegionOne',                                                 <- not from user. 
          name: <string>(adc.id + '-' + net.type + '-' + k),
        };
```

Reply:
what if region != "RegionOne" in an  actual environment?

Reply:
As we talked f2f, putting into configuration file/environment is reasonable. 
Prefer to do it in a separate PR.